src/vivarium_census_prl_synth_pop/components/observers.py

```diff
@@ -165,3 +166,138 @@ def do_observation(self, event) -> None:
+        pop["wic_year"] = event.time.year
+        pop["middle_initial"] = pop["middle_name"].astype(str).str[0]
+        pop = pop.drop(columns="middle_name")
+
+        # merge address ids for guardian_1 and guardian_2 for the rows with guardians
+        for i in [1,2]:
+            s_guardian_id = pop[f"guardian_{i}"].dropna()
+            s_guardian_id = s_guardian_id[s_guardian_id != -1] # is it faster to remove the negative values?
+            pop[f"guardian_{i}_address_id"] = s_guardian_id.map(pop["address_id"])
+
+        pop["census_year"] = event.time.year
+
+
+        # add addl columns for simulating coverage
+        pop["nominal_age"] = np.floor(pop["age"])
+        
+        # calculate household size and income for measuring WIC eligibility
+        hh_size = pop.address_id.value_counts()
+        pop["hh_size"] = pop.address_id.map(hh_size)
+
+        hh_income = pop.groupby("address_id").income.sum()
+        pop["hh_income"] = pop.address_id.map(hh_income)
+
+        
+        # income eligibility for WIC is total household income less
+        # than $16,410 + ($8,732 * number of people in the household)
+        pop["wic_eligible"] = pop.eval("hh_income <= 16_410 + 8_732*hh_size")
+        
+        
+        # filter population to mothers and children under 5
+        pop_u1 = pop.query("age < 1 and wic_eligible")
+        pop_1_to_5 = pop.query("(age >= 1) and (age < 5) and wic_eligible")
+
+        guardian_ids = set(pop_u1["guardian_1"]) | set(pop_u1["guardian_2"])
+        pop_mothers = pop[(pop["sex"] == "Female") & pop.index.isin(guardian_ids) & pop.wic_eligible]
+
+
+        # determine who is covered using age/race-specific coverage probabilities
+        # with additional constraint that all under-1 year olds with mother covered are also covered
+
+        # first include some mothers
+        pr_covered = data_values.COVERAGE_PROBABILITY_WIC['mothers']
+        mother_covered_probability = pop_mothers.race_ethnicity.map(pr_covered)
+        pop_included_mothers = self.randomness.filter_for_probability(pop_mothers, mother_covered_probability)
+
+        # then use same pattern for children aged 1 to 4
+        pop_included = {}
+        for age, pop_age in pop_1_to_5.groupby("nominal_age"):
+            pr_covered = data_values.COVERAGE_PROBABILITY_WIC[age]
+            child_covered_pr = pop_age.race_ethnicity.map(pr_covered)
+            pop_included[age] = self.randomness.filter_for_probability(pop_age, child_covered_pr)
+
+
+        # selection for age 0 is more complicated; it should include
+        # all simulants who have a mother enrolled and then a random
+        # selection of additional simulants to reach the covered
+        # probabilities
+        child_covered_pr = (pop_u1.guardian_1.isin(pop_included_mothers.index)
+                            | pop_u1.guardian_2.isin(pop_included_mothers.index)).astype(float)
+        for race_eth in ["Latino", "Black", "White", "Other"]:
```

Comment:
If you follow my comments below regarding the default dict, you can remove a lot of this logic and just use the `simplified_race_ethnicity` Series you would have created here.